// grommet-controls/dist/components/VerticalMenu/VerticalMenu.stories.js
import React from 'react';
import { Box, Text } from 'grommet';
import { Inbox, Send, Flag, Star, Document, Tag, Trash } from 'grommet-icons';
import { VerticalMenu } from './VerticalMenu';
export default {
    title: 'navigation/VerticalMenu',
    component: VerticalMenu,
};
export const main = () => (React.createElement(VerticalMenu, { activeItem: { id: 'c_drive' }, items: [
        {
            id: 'local',
            label: 'Local',
            items: [
                { id: 'c_drive', href: '/c_drive', label: 'C: drive' },
                {
                    id: 'f_drive',
                    onClick: () => console.log('onClick'),
                    label: 'F: drive',
                },
            ],
        },
        {
            id: 'cloud',
            label: 'Cloud',
            items: [
                { id: 'drop_box', href: '/drop_box', label: 'DropBox' },
                {
                    id: 'google_drive',
                    onClick: () => console.log('onClick'),
                    label: 'Google drive',
                },
            ],
        },
    ] }));
const Badge = ({ label, background = 'light-3' }) => (React.createElement(Box, { background: background, pad: { horizontal: 'small' }, border: "all", style: { borderRadius: '20%' } },
    React.createElement(Text, { size: "small" }, label)));
export const items = () => (React.createElement(VerticalMenu, { activeItem: { id: 'all' }, items: [
        {
            id: 'inbox',
            label: 'Inbox',
            icon: React.createElement(Inbox, null),
            expanded: true,
            items: [
                {
                    id: 'all',
                    href: '/inbox/all',
                    label: 'All',
                    widget: React.createElement(Badge, { label: "10" }),
                },
                {
                    id: 'gmail',
                    href: '/inbox/gmail',
                    label: 'GMail',
                    widget: React.createElement(Badge, { label: "8" }),
                },
                {
                    id: 'work',
                    href: '/inbox/work',
                    label: 'Work',
                    widget: React.createElement(Badge, { label: "1" }),
                },
                {
                    id: 'amazon',
                    href: '/inbox/amazon',
                    label: 'Amazon',
                    widget: React.createElement(Badge, { label: "1" }),
                },
            ],
        },
        {
            id: 'sent',
            href: '/inbox/sent',
            label: 'Sent',
            icon: React.createElement(Send, null),
        },
        {
            id: 'flagged',
            href: '/inbox/flagged',
            label: 'Flagged',
            icon: React.createElement(Flag, null),
            widget: React.createElement(Badge, { label: "3", background: "accent-1" }),
        },
        {
            id: 'starred',
            href: '/inbox/starred',
            label: 'Starred',
            icon: React.createElement(Star, null),
        },
        {
            id: 'drafts',
            href: '/inbox/drafts',
            label: 'Drafts',
            icon: React.createElement(Document, null),
        },
        {
            id: 'tagged',
            href: '/inbox/tagged',
            label: 'Tagged',
            icon: React.createElement(Tag, null),
        },
        {
            id: 'trash',
            href: '/inbox/trash',
            label: 'Trash',
            icon: React.createElement(Trash, null),
        },
    ] }));
