// grommet-controls/dist/components/Value/Value.stories.js
import React from 'react';
import { Box, Text } from 'grommet';
import { Down } from 'grommet-icons';
import { Value } from './Value';
export default {
    title: 'controls/Value',
    component: Value,
};
export const main = props => (React.createElement(Box, { direction: "row" },
    React.createElement(Box, { basis: "small" },
        React.createElement(Value, Object.assign({}, props)))));
main.controls = {
    value: '30%',
    label: 'last quarter sales',
};
export const label = () => React.createElement(Value, { value: "30%", label: "a descriptive label" });
export const value = () => (React.createElement(Value, { value: React.createElement(Box, { direction: "row", align: "center", gap: "xsmall" },
        React.createElement(Text, { size: "large", weight: "bold", color: "status-error" }, "30%"),
        React.createElement(Down, { color: "status-error" })), label: "custom value" }));
export const color = () => (React.createElement(Value, { label: "status-ok", value: "30%", color: "status-ok" }));
export const weight = () => (React.createElement(Box, { direction: "row" },
    React.createElement(Box, { basis: "small", gap: "small" },
        React.createElement(Value, { weight: "normal", value: "30%", label: "normal" }),
        React.createElement(Value, { weight: "bold", value: "30%", label: "bold" }),
        React.createElement(Value, { weight: 400, value: "30%", label: "400" }))));
export const size = () => (React.createElement(Box, { direction: "row" },
    React.createElement(Box, { basis: "small", gap: "small" },
        React.createElement(Value, { size: "xsmall", value: "30%", label: "xsmall" }),
        React.createElement(Value, { size: "small", value: "30%", label: "small" }),
        React.createElement(Value, { size: "medium", value: "30%", label: "medium" }),
        React.createElement(Value, { size: "large", value: "30%", label: "large" }),
        React.createElement(Value, { size: "xlarge", value: "30%", label: "xlarge" }),
        React.createElement(Value, { size: "xxlarge", value: "30%", label: "xxlarge" }))));
