// grommet-controls/dist/components/Form/Form.stories.js
import React from 'react';
import { Box, Button, Text } from 'grommet';
import { Form } from './Form';
import { TextInputField, PasswordInputField, SelectField, CheckBoxField, NumberInputField, validators, } from '../../index';
export default {
    title: 'validation/Form',
    component: Form,
};
export const main = () => (React.createElement(Form, { onSubmit: () => console.log('onSubmit'), pad: {
        horizontal: 'small',
    } },
    React.createElement(PasswordInputField, { label: "Password", name: "password", validation: [
            validators.required(),
            validators.minLength(8),
            validators.alphaNumeric(),
        ] }),
    React.createElement(PasswordInputField, { label: "Confirm Password", name: "password1", validation: [validators.equalsField('password', 'the above password')] }),
    React.createElement(TextInputField, { label: "URL", name: "url", validation: [validators.required(), validators.url()] }),
    React.createElement(SelectField, { name: "gender", options: ['male', 'female'], validation: [validators.required()] }),
    React.createElement(CheckBoxField, { name: "tos", label: "Terms of service", validation: [
            validators.required(),
            validators.True('Please accept the TOS'),
        ] }),
    React.createElement(NumberInputField, { min: 6, max: 22, name: "age", label: "Age", validation: [
            validators.required(),
            validators.numeric(),
            validators.bigger(10),
            validators.smallerOrEqual(18),
        ] }),
    React.createElement(Box, { pad: "small" },
        React.createElement(Button, { type: "submit", label: "Submit" }))));
export const onSubmit = () => (React.createElement(Form, { focusFirstChild: false, onSubmit: () => console.log('onSubmit'), basis: "small" },
    React.createElement(TextInputField, { label: "Text", name: "fieldname" })));
export const onChange = () => {
    const [changedValue, setChangedValue] = React.useState('');
    return (React.createElement(Box, { pad: "small" },
        React.createElement(Form, { focusFirstChild: false, onChange: (name, value) => {
                console.log(name, value);
                setChangedValue(value);
            }, basis: "small" },
            React.createElement(TextInputField, { label: "Text", name: "mytextinput" })),
        React.createElement(Text, { size: "small" }, changedValue)));
};
export const onSubmitError = () => (React.createElement(Box, { pad: "small" },
    React.createElement(Form, { focusFirstChild: false, onSubmit: () => console.log('onSubmit'), onSubmitError: () => console.log('onSubmitError'), basis: "small" },
        React.createElement(TextInputField, { label: "Text", name: "errofield", validation: [validators.required(), validators.minLength(8)] }))));
export const onInvalidForm = () => {
    const [invalid, setInvalid] = React.useState(undefined);
    return (React.createElement(Box, { pad: "small" },
        React.createElement(Form, { focusFirstChild: false, onSubmit: () => console.log('onSubmit'), onInvalidForm: error => setInvalid(error), onValidForm: () => setInvalid(undefined), basis: "small" },
            React.createElement(TextInputField, { label: "Text", name: "invalidfield", validation: [validators.required(), validators.minLength(8)] })),
        React.createElement(Text, { size: "small" }, invalid ? 'invalid' : 'valid')));
};
