// grommet-controls/dist/components/ImageStamp/ImageStamp.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { StyledImageStamp } from './StyledImageStamp';
const SIZE_MAP = {
    small: 'medium',
    medium: 'large',
    large: 'xlarge',
    xlarge: 'xxlarge',
};
/**
 *  A small image stamp used for avatars or image thumbnails
 */
const ImageStamp = (_a) => {
    var { size, src } = _a, rest = __rest(_a, ["size", "src"]);
    return src ? React.createElement(StyledImageStamp, Object.assign({ src: src, size: SIZE_MAP[size] }, rest)) : null;
};
ImageStamp.defaultProps = {
    contain: null,
    size: 'medium',
    plain: false,
};
export { ImageStamp };
