// grommet-controls/dist/components/DropInput/DropInput.stories.js
import React from 'react';
import { Box, Calendar } from 'grommet';
import { Calendar as CalendarIcon, Add, Subtract } from 'grommet-icons';
import { DropInput } from './DropInput';
import { smallDate } from '../../utils/moment';
export default {
    title: 'input/DropInput',
    component: DropInput,
};
const DateInput = props => {
    const [date, setDate] = React.useState(smallDate(new Date('08/30/2020')));
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium" },
            React.createElement(DropInput, Object.assign({ dropContent: React.createElement(Box, { pad: "small" },
                    React.createElement(Calendar, { size: "small", date: date, onSelect: isoDate => setDate(smallDate(new Date(isoDate))) })), value: date, onChange: ({ target }) => setDate(target.value) }, props)))));
};
const NumberInput = () => {
    const [number, setNumber] = React.useState(10);
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium" },
            React.createElement(DropInput, { value: number, onChange: ({ target: { value } }) => setNumber(parseFloat(value)), widgets: [
                    {
                        icon: React.createElement(Add, null),
                        onClick: () => setNumber(number + 1),
                    },
                    {
                        icon: React.createElement(Subtract, null),
                        onClick: () => setNumber(number - 1),
                    },
                ] }))));
};
const DropInputTest = props => {
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium" },
            React.createElement(DropInput, Object.assign({}, props)))));
};
export const main = () => React.createElement(DateInput, null);
export const a11yTitle = () => React.createElement(DateInput, { a11yTitle: "Birthdy date" });
export const disabled = () => React.createElement(DateInput, { disabled: true });
export const dropContent = () => (React.createElement(DropInputTest, { a11yDropTitle: "Open calendar", dropContent: React.createElement(Box, { pad: "small" },
        React.createElement(Calendar, { size: "small" })), defaultValue: smallDate(new Date('08/30/2020')) }));
export const dropIcon = () => React.createElement(DateInput, { dropIcon: React.createElement(CalendarIcon, null) });
export const widgets = () => React.createElement(NumberInput, null);
export const plain = () => (React.createElement(DropInputTest, { plain: true, defaultValue: "Plain no focus" }));
export const focusIndicator = () => (React.createElement(DropInputTest, { plain: true, focusIndicator: true, defaultValue: "Plain with focus" }));
export const placeholder = () => React.createElement(DropInputTest, { placeholder: "Enter phone..." });
export const name = () => React.createElement(DropInputTest, { id: "date-id", name: "date-name" });
