// grommet-controls/dist/components/DropInput/DropInput.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import { ThemeContext } from 'styled-components';
import { FormDown } from 'grommet-icons/icons/FormDown';
import { DropButton, Keyboard, Button } from 'grommet';
import { StyledDropInput, StyledDropInputContainer, StyledWidgetsContainer, } from './StyledDropInput';
/**
 *  An Input control with an optional drop button with the specified 'dropContent' or widgets<br/>
 *`import { DropInput } from 'grommet-controls';`<br/>
 *`<DropInput`<br/>
 *&nbsp;&nbsp;`dropContent={(`<br/>
 *&nbsp;&nbsp;&nbsp;&nbsp;`...`<br/>
 *&nbsp;&nbsp;`]}`<br/>
 *`/>`<br/>
 */
class DropInput extends Component {
    constructor() {
        super(...arguments);
        this.state = { open: false };
        this.inputRef = React.createRef();
        this.onUpdateValue = value => {
            const { onChange } = this.props;
            // eslint-disable-next-line react/no-find-dom-node
            const input = findDOMNode(this.inputRef.current);
            input.value = value;
            this.setState({ open: false, value });
            if (onChange) {
                onChange({ target: input });
            }
        };
        this.onOpen = e => {
            const { onOpen, disabled } = this.props;
            this.setState({ open: true });
            if (onOpen && !disabled) {
                onOpen(e);
            }
        };
        this.onClose = e => {
            const { onClose, disabled } = this.props;
            this.setState({ open: false });
            if (onClose && !disabled) {
                onClose(e);
            }
        };
        this.onKeyDown = e => {
            const { onKeyDown } = this.props;
            this.setState({ open: true });
            if (onKeyDown) {
                onKeyDown(e);
            }
        };
        this.onKeyUp = e => {
            const { onKeyUp } = this.props;
            this.setState({ open: false });
            if (onKeyUp) {
                onKeyUp(e);
            }
        };
    }
    static getDerivedStateFromProps(newProps, oldState) {
        if (newProps.value !== oldState.value ||
            newProps.defaultValue !== oldState.defaultValue) {
            return {
                value: newProps.value,
                defaultValue: newProps.defaultValue,
                open: false,
            };
        }
        return null;
    }
    render() {
        const _a = this.props, { a11yTitle, a11yDropTitle, dropAlign, dropTarget, update, widgets, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onOpen, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onClose, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onKeyDown, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onKeyUp, defaultValue, dropContent, dropIcon, disabled } = _a, rest = __rest(_a, ["a11yTitle", "a11yDropTitle", "dropAlign", "dropTarget", "update", "widgets", "onOpen", "onClose", "onKeyDown", "onKeyUp", "defaultValue", "dropContent", "dropIcon", "disabled"]);
        const { open } = this.state;
        if (typeof update === 'function') {
            update(this.onUpdateValue);
        }
        const numWidgets = (dropContent ? 1 : 0) + (widgets ? widgets.length : 0);
        let decorations;
        if (numWidgets > 0) {
            let drop;
            if (dropContent) {
                drop = (React.createElement(DropButton, { a11yTitle: a11yDropTitle, disabled: disabled, dropAlign: dropAlign, dropTarget: dropTarget, open: open, tabIndex: -1, focusIndicator: false, onOpen: this.onOpen, onClose: this.onClose, dropContent: dropContent, icon: dropIcon }));
            }
            decorations = (React.createElement(StyledWidgetsContainer, { align: "center", direction: "row" },
                widgets.map((widget, index) => (React.createElement(Button, Object.assign({ disabled: disabled, tabIndex: -1, key: `widget_${index}` }, widget)))),
                drop));
        }
        return (React.createElement(Keyboard, { onDown: this.onKeyDown, onUp: this.onKeyUp },
            React.createElement(StyledDropInputContainer, null,
                React.createElement(ThemeContext.Consumer, null, theme => (React.createElement(StyledDropInput, Object.assign({ ref: this.inputRef, theme: theme, disabled: disabled, numWidgets: numWidgets, "aria-label": a11yTitle, defaultValue: defaultValue ? defaultValue.toString() : undefined }, rest)))),
                decorations)));
    }
}
DropInput.defaultProps = {
    dropAlign: { top: 'bottom', right: 'left' },
    dropIcon: React.createElement(FormDown, null),
    type: 'text',
    widgets: [],
};
export { DropInput };
