// grommet-controls/dist/components/PagingTable/components/Pagination.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { Component } from 'react';
import styled from 'styled-components';
import { Box, Button, Select, Text } from 'grommet';
import { Previous } from 'grommet-icons/icons/Previous';
import { Next } from 'grommet-icons/icons/Next';
import { NumberInput } from '../../NumberInput';
const defaultButton = (_a) => {
    var { disabled, Icon, onClick, label } = _a, other = __rest(_a, ["disabled", "Icon", "onClick", "label"]);
    return (React.createElement(Button, Object.assign({ icon: React.createElement(Icon, null), disabled: disabled, onClick: onClick, label: label, plain: true }, other)));
};
const StyledButton = styled(defaultButton) `
  width: 150px;
  padding: 10px 0;
`;
const StyledPageInput = styled(NumberInput) `
  max-width: 180px;
`;
const StyledSelect = styled(Select) `
  max-width: 150px;
`;
export default class ReactTablePagination extends Component {
    constructor(props, context) {
        super(props, context);
        this.getSafePage = page => {
            const pg = Number.isNaN(page) ? this.props.page : page;
            return Math.min(Math.max(pg, 0), this.props.pages - 1);
        };
        this.changePage = page => {
            const pg = this.getSafePage(page);
            this.setState({ page: pg });
            if (this.props.page !== pg) {
                this.props.onPageChange(pg);
            }
        };
        this.applyPage = e => {
            if (e) {
                e.preventDefault();
            }
            const { page } = this.state;
            this.changePage(page === '' ? this.props.page : page);
        };
        this.state = { page: props.page };
    }
    static getDerivedStateFromProps(nextProps) {
        return {
            page: nextProps.page,
        };
    }
    renderPaging() {
        const { page, showPageJump, pageText, ofText, pages } = this.props;
        let pageJump;
        if (showPageJump) {
            pageJump = (
            //@ts-ignore
            React.createElement(StyledPageInput, { "aria-label": "Select page to jump to", onChange: e => {
                    const val = e.target.value;
                    if (val === '') {
                        return this.setState({ page: val });
                    }
                    const pg = parseInt(val, 10) - 1;
                    this.setState({ page: this.getSafePage(pg) });
                    return false;
                }, value: typeof this.state.page === 'string' ? '' : this.state.page + 1, onBlur: this.applyPage, onKeyPress: e => {
                    if (e.which === 13 || e.keyCode === 13) {
                        this.applyPage(null);
                    }
                } }));
        }
        else {
            pageJump = React.createElement(Text, null, page + 1);
        }
        return (React.createElement(Box, { direction: "row", align: "center", gap: "small" },
            React.createElement(Text, null, `${pageText} `),
            pageJump,
            React.createElement("span", { style: { whiteSpace: 'nowrap' } }, `${ofText} ${pages ||
                1}`)));
    }
    renderPageSize() {
        const { showPageSizeOptions, onPageSizeChange, pageSizeOptions, pageSize, } = this.props;
        if (showPageSizeOptions) {
            return (React.createElement(Box, { direction: "row" },
                React.createElement(StyledSelect, { "aria-label": "Select rows per page", onChange: (e) => onPageSizeChange(Number(e.option.split(' ')[0])), value: `${pageSize} ${this.props.rowsText}`, options: pageSizeOptions.map(option => `${option} ${this.props.rowsText}`) })));
        }
        return null;
    }
    renderPrevious() {
        const { PreviousComponent = StyledButton, canPrevious, page } = this.props;
        return (React.createElement(Box, { direction: "row" },
            React.createElement(PreviousComponent, { "aria-label": "Move to previous page", Icon: Previous, label: this.props.previousText, onClick: canPrevious ? () => this.changePage(page - 1) : null, disabled: !canPrevious })));
    }
    renderNext() {
        const { page, canNext, NextComponent = StyledButton } = this.props;
        return (React.createElement(Box, { direction: "row" },
            React.createElement(NextComponent, { "aria-label": "Move to next page", reverse: true, Icon: Next, label: this.props.nextText, onClick: canNext ? () => this.changePage(page + 1) : null, disabled: !canNext })));
    }
    render() {
        const props = this.props.getPaginationProps(this.props);
        return (React.createElement(Box, Object.assign({}, Object.assign({
            direction: 'row',
            align: 'center',
            justify: 'between',
            fill: 'horizontal',
        }, props)),
            this.renderPrevious(),
            this.renderPaging(),
            this.renderPageSize(),
            this.renderNext()));
    }
}
