// grommet-controls/dist/components/PagingTable/components/TrComponent.js
import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import { Keyboard } from 'grommet';
import { focusedChildIndex, focusChildByIndex, focusNextElement, focusPrevElement, hasKeyboardChildren, } from '../DOM';
import { StyledTrComponent } from '../StyledTableComponents';
class TrComponent extends Component {
    constructor() {
        super(...arguments);
        this.selectNextRow = () => {
            const { selectRow } = this.props;
            const activeChild = focusedChildIndex(this.container);
            return selectRow(true, activeChild);
        };
        this.selectPrevRow = () => {
            const { selectRow } = this.props;
            const activeChild = focusedChildIndex(this.container);
            return selectRow(false, activeChild);
        };
        this.selectNextChild = () => {
            if (focusNextElement(this.container)) {
                return true;
            }
            return this.selectNextRow();
        };
        this.selectPrevChild = () => {
            if (focusPrevElement(this.container)) {
                return true;
            }
            return this.selectPrevRow();
        };
        this.selectFirstChild = activeIndex => focusChildByIndex(this.container, activeIndex);
    }
    render() {
        return (React.createElement(Keyboard, { onTab: e => {
                e.preventDefault();
                if (e.shiftKey) {
                    this.selectPrevChild();
                }
                else {
                    this.selectNextChild();
                }
            }, onLeft: e => {
                if (!hasKeyboardChildren(document.activeElement)) {
                    e.preventDefault();
                    this.selectPrevChild();
                }
            }, onRight: e => {
                if (!hasKeyboardChildren(document.activeElement)) {
                    e.preventDefault();
                    this.selectNextChild();
                }
            }, onUp: e => {
                if (!hasKeyboardChildren(document.activeElement)) {
                    e.preventDefault();
                    this.selectPrevRow();
                }
            }, onDown: e => {
                if (!hasKeyboardChildren(document.activeElement)) {
                    e.preventDefault();
                    this.selectNextRow();
                }
            } },
            React.createElement(StyledTrComponent, Object.assign({ direction: "row", role: "row", ref: el => {
                    // eslint-disable-next-line react/no-find-dom-node
                    this.container = findDOMNode(el);
                } }, this.props))));
    }
}
export default TrComponent;
