// grommet-controls/dist/components/Notification/Notification.stories.js
import React from 'react';
import { Grommet, CircleQuestion } from 'grommet-icons';
import { ControlTypes } from '@component-controls/core';
import { Notification } from './Notification';
export default {
    title: 'controls/Notification',
    component: Notification,
};
const onClose = () => console.log('Close clicked');
export const main = props => React.createElement(Notification, Object.assign({}, props));
main.controls = {
    border: {
        type: ControlTypes.OBJECT,
        value: {
            color: {
                type: ControlTypes.COLOR,
                value: 'brand',
            },
            size: {
                type: ControlTypes.OPTIONS,
                options: ['xsmall', 'small', 'medium', 'large', 'xlarge'],
                value: 'medium',
            },
            side: {
                type: ControlTypes.OPTIONS,
                options: [
                    'top',
                    'left',
                    'bottom',
                    'right',
                    'horizontal',
                    'vertical',
                    'all',
                ],
                value: 'all',
            },
        },
    },
    message: { type: ControlTypes.TEXT, value: 'Message heading' },
    state: { type: ControlTypes.TEXT, value: 'state label' },
    timestamp: {
        type: ControlTypes.DATE,
        value: new Date('Sun, Aug 30, 2020, 11:14 PM'),
    },
    strong: { type: ControlTypes.BOOLEAN, value: true },
    percentComplete: {
        type: ControlTypes.NUMBER,
        value: 30,
        min: 0,
        max: 100,
        range: true,
    },
    status: {
        type: ControlTypes.OPTIONS,
        options: ['ok', 'info', 'warning', 'error', 'unknown', 'disabled'],
        value: 'ok',
    },
    onClose: { type: ControlTypes.BUTTON, value: onClose },
};
export const a11yTitle = () => (React.createElement(Notification, { a11yTitle: "Close notification", onClose: onClose }));
export const status = () => (React.createElement(Notification, { message: "Message heading", status: "info", onClose: onClose }));
export const state = () => (React.createElement(Notification, { message: "Message heading", state: "state note", status: "warning", size: "small", onClose: onClose }));
export const strong = () => (React.createElement(Notification, { message: "Bold message", strong: true, status: "error", size: "small", onClose: onClose }));
export const icon = () => (React.createElement(Notification, { message: "Custom icon", icon: React.createElement(Grommet, null), size: "small", onClose: onClose }));
export const closer = () => (React.createElement(Notification, { status: "disabled", message: "Custom closer", closer: React.createElement(CircleQuestion, null), size: "small", onClose: onClose }));
export const percentComplete = () => (React.createElement(Notification, { status: "info", message: "Task percentage", percentComplete: 80, size: "small", onClose: onClose }));
export const timestamp = () => (React.createElement(Notification, { message: "Time stamp", timestamp: new Date('Sun, Aug 30, 2020, 11:14 PM'), size: "small", onClose: onClose }));
export const locale = () => (React.createElement(Notification, { message: "Locale de-DE", timestamp: new Date('Sun, Aug 30, 2020, 11:14 PM'), locale: "de-DE", size: "small", onClose: onClose }));
export const size = () => (React.createElement(Notification, { message: "Large", size: "large", onClose: onClose }));
export const reverse = () => (React.createElement(Notification, { message: "reversed", size: "small", reverse: true, onClose: onClose }));
