// grommet-controls/dist/components/WithFormField/WithFormField.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { useEffect, useContext } from 'react';
import { FormField } from 'grommet';
import { FormContext } from '../Form/Form';
export const WithFormField = (_a) => {
    var { validation, label, description, name, controlLabel, inField, children } = _a, other = __rest(_a, ["validation", "label", "description", "name", "controlLabel", "inField", "children"]);
    const form = useContext(FormContext);
    const { onFieldChange, getFieldValue, getFieldErrors } = form;
    useEffect(() => {
        form.attachToForm(name, { label, description, validation });
        return () => {
            form.detachFromForm(name);
        };
    });
    const onChange = e => {
        const { value, target } = e;
        const targetValue = target ? target.value : undefined;
        onFieldChange(name, value || targetValue, e);
    };
    const getValue = () => {
        const value = getFieldValue(name);
        return value === undefined ? '' : value;
    };
    const errors = getFieldErrors(name);
    let error = Array.isArray(errors) ? errors[0] : errors;
    if (typeof error === 'function') {
        error = error(typeof label === 'string' ? label : name);
    }
    const field = children(Object.assign({ id: name, name, label: inField ? controlLabel : label, value: getValue(), onChange }, other));
    if (!inField) {
        return field;
    }
    return (React.createElement(FormField, { htmlFor: name, label: label || (controlLabel === true && name), error: error }, field));
};
WithFormField.defaultProps = {
    validation: undefined,
    inField: true,
    controlLabel: undefined,
    label: undefined,
    description: undefined,
    Control: React.Component,
};
