// grommet-controls/dist/components/DateInput/DateInput.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { Component } from 'react';
import { Box, Calendar } from 'grommet';
import { Calendar as CalendarIcon } from 'grommet-icons/icons/Calendar';
import { MaskedInput } from '../MaskedInput';
import { smallDate } from '../../utils/moment';
/**
 * A masked date entry field with a drop down calendar<br/>
 * All properties of MaskedInput apply<br/>
 * `import { DateInput } from 'grommet-controls';`<br/>
 * `<DateInput value={...} />`<br/>
 */
class DateInput extends Component {
    constructor() {
        super(...arguments);
        this.upDateValue = null;
        this.onSelect = isoDate => {
            const date = new Date(isoDate);
            this.upDateValue(smallDate(date));
        };
    }
    render() {
        const _a = this.props, { value, bounds, dates, disabledDates, autocorrect, firstDayOfWeek, locale, size, mask: userMask, maskFormat } = _a, rest = __rest(_a, ["value", "bounds", "dates", "disabledDates", "autocorrect", "firstDayOfWeek", "locale", "size", "mask", "maskFormat"]);
        let mask;
        if (userMask) {
            mask = { mask: userMask };
        }
        else if (autocorrect) {
            mask = {
                mask: [/\d/, /\d/, '/', /\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/],
                pipe: MaskedInput.createAutoCorrectedDatePipe(maskFormat),
                keepCharPositions: true,
            };
        }
        else {
            mask = {
                mask: [/\d/, /\d/, '/', /\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/],
            };
        }
        const controlledValue = value === undefined ? '' : value;
        return (React.createElement(MaskedInput, Object.assign({ update: update => {
                this.upDateValue = update;
            }, value: controlledValue, dropContent: React.createElement(Box, { pad: "small" },
                React.createElement(Calendar, { date: controlledValue.toString(), onSelect: this.onSelect, bounds: bounds, dates: dates, disabled: disabledDates, firstDayOfWeek: firstDayOfWeek, locale: locale, size: size })) }, Object.assign(Object.assign({}, rest), mask))));
    }
}
DateInput.defaultProps = {
    dropIcon: React.createElement(CalendarIcon, null),
    size: 'small',
    a11yDropTitle: 'Open calendar',
};
export { DateInput };
