// grommet-controls/dist/components/DateInput/DateInput.stories.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { Box } from 'grommet';
import { DateInput } from './DateInput';
import { smallDate } from '../../utils/moment';
export default {
    title: 'input/DateInput',
    component: DateInput,
};
const DateTestBed = (_a) => {
    var { value } = _a, rest = __rest(_a, ["value"]);
    const [stateValue, setValue] = React.useState(value);
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium" },
            React.createElement(DateInput, Object.assign({ value: stateValue, onChange: ({ target }) => setValue(target.value) }, rest)))));
};
export const main = () => (React.createElement(DateTestBed, { value: smallDate(new Date('08/30/2020')) }));
export const autocorrect = () => (React.createElement(DateTestBed, { value: smallDate(new Date('08/30/2020')), autocorrect: true }));
export const disabled = () => (React.createElement(DateTestBed, { value: smallDate(new Date('08/30/2020')), disabled: true }));
export const bounds = () => (React.createElement(DateTestBed, { value: smallDate(new Date('08/30/2020')), bounds: [
        smallDate(new Date(new Date('08/30/2020').getFullYear(), 0, 1)),
        smallDate(new Date(new Date('08/30/2020').getFullYear(), 11, 31)),
    ] }));
export const disabledDates = () => (React.createElement(DateTestBed, { value: smallDate(new Date('08/30/2020')), disabledDates: [
        smallDate(new Date('08/30/2020')),
        smallDate(new Date('08/30/2020').setDate(new Date('08/30/2020').getDate() - 1)),
    ] }));
export const firstDayOfWeek = () => (React.createElement(DateTestBed, { value: smallDate(new Date('08/30/2020')), firstDayOfWeek: 1 }));
export const locale = () => (React.createElement(DateTestBed, { value: smallDate(new Date('08/30/2020')), locale: "de-DE" }));
export const size = () => (React.createElement(DateTestBed, { value: smallDate(new Date('08/30/2020')), size: "medium" }));
