// grommet-controls/dist/components/Tags/Tags.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import { Box, Keyboard, Text, } from 'grommet';
import { FormClose } from 'grommet-icons/icons/FormClose';
import { Tag } from '../Tag';
export const TagsSelect = (props) => {
    const _a = props || { tagProps: {} }, { tagProps } = _a, rest = __rest(_a, ["tagProps"]);
    return ({ placeholder, value, onChange }) => (React.createElement(Tags, Object.assign({ focusable: false, placeholder: placeholder, value: value, onChange: onChange, tagProps: Object.assign(Object.assign({}, tagProps), { onClick: e => e.stopPropagation() }) }, rest)));
};
/**
 * A list of tags that can be removed<b/>
 * `import { Tags } from 'grommet-controls';`<b/>
 * `<Tags />`<b/>
 */
class Tags extends Component {
    constructor() {
        super(...arguments);
        this.state = {
            selectedTagIndex: -1,
        };
        this.tagRefs = [];
        this.elementRef = React.createRef();
        this.selectTag = (selected) => {
            const { onChange, value } = this.props;
            if (onChange) {
                let tags = selected;
                if (Array.isArray(value)) {
                    const index = value.indexOf(tags);
                    if (index !== -1) {
                        tags = value.filter(item => item !== selected);
                    }
                    else {
                        tags = [...value, tags];
                    }
                }
                else {
                    tags = [tags];
                }
                onChange({ target: findDOMNode(this.elementRef.current), option: selected, value: tags });
            }
        };
        this.onNextTag = (event) => {
            const { value } = this.props;
            const { selectedTagIndex } = this.state;
            event.preventDefault();
            let index = selectedTagIndex + 1;
            if (index >= value.length) {
                index = 0;
            }
            this.focusTag(index);
        };
        this.onPreviousTag = (event) => {
            const { selectedTagIndex } = this.state;
            const { value } = this.props;
            event.preventDefault();
            let index = selectedTagIndex - 1;
            if (index < 0) {
                index = value.length - 1;
            }
            this.focusTag(index);
        };
        this.onSelectTag = (event) => {
            const { value } = this.props;
            const { selectedTagIndex } = this.state;
            if (selectedTagIndex >= 0 && selectedTagIndex < value.length) {
                event.preventDefault(); // prevent submitting forms
                event.stopPropagation();
                this.selectTag(value[selectedTagIndex]);
            }
        };
        this.onCloseClick = (e, tag) => {
            e.stopPropagation();
            this.selectTag(tag);
        };
    }
    focusTag(index) {
        if (index >= 0 && index < this.tagRefs.length) {
            const tagElement = findDOMNode(this.tagRefs[index].current);
            if (tagElement && typeof tagElement.focus === 'function') {
                tagElement.focus();
            }
            this.setState({ selectedTagIndex: index });
        }
    }
    render() {
        const _a = this.props, { placeholder, children, value, onChange, focusable, onClick, direction, icon, tagProps } = _a, rest = __rest(_a, ["placeholder", "children", "value", "onChange", "focusable", "onClick", "direction", "icon", "tagProps"]);
        let noValues;
        if ((!value || (Array.isArray(value) && value.length === 0))) {
            noValues = React.isValidElement(placeholder) ? placeholder : (
            // placeholder. minimum height of icon to keep size
            React.createElement(Text, { color: 'placeholder', style: { minHeight: '24px' } }, placeholder || 'No selection'));
        }
        let values;
        if (Array.isArray(value)) {
            values = value;
        }
        else {
            values = value ? [value] : [];
        }
        return (React.createElement(Keyboard, { onEnter: this.onSelectTag, onSpace: this.onSelectTag, onLeft: this.onPreviousTag, onRight: this.onNextTag },
            React.createElement(Box, Object.assign({ tabIndex: focusable ? 0 : undefined, ref: this.elementRef, direction: direction, overflow: 'auto', style: { minWidth: 'auto' } }, rest), noValues || values.map((tag, index) => {
                if (children) {
                    return children(tag, index, value);
                }
                if (!this.tagRefs[index]) {
                    this.tagRefs[index] = React.createRef();
                }
                return (React.createElement(Tag, Object.assign({ key: `tag_${tag}_${index}`, ariaChecked: true, a11yTitle: `Remove ${tag.toString()}`, label: typeof tag !== 'object' ? tag.toString() : undefined, ref: this.tagRefs[index], onClick: onClick ? e => onClick(e, tag) : undefined, onChange: onChange ? e => this.onCloseClick(e, tag) : undefined, icon: icon }, (typeof tag === 'object' ? Object.assign(Object.assign({}, tagProps), tag) : tagProps))));
            }))));
    }
}
Tags.defaultProps = {
    pad: 'small',
    focusable: true,
    margin: { horizontal: 'xsmall', vertical: 'small' },
    value: [],
    gap: 'xsmall',
    direction: 'row-responsive',
    icon: React.createElement(FormClose, null),
};
export { Tags };
