// grommet-controls/dist/components/Tags/Tags.stories.js
import React from 'react';
import { Box, Button } from 'grommet';
import { FormSubtract, Trash } from 'grommet-icons';
import { Tags } from './Tags';
export default {
    title: 'input/Tags',
    component: Tags,
};
const stringOptions = ['small', 'medium', 'large', 'xlarge', 'huge'];
const defaultTags = [stringOptions[0], stringOptions[2]];
const TagsInput = props => {
    const [tags, setTags] = React.useState(defaultTags);
    const onChangeTags = ({ option }) => setTags(option);
    const resetTags = () => setTags(defaultTags);
    return (React.createElement(Box, { gap: "large" },
        React.createElement(Box, { direction: "row" },
            React.createElement(Tags, Object.assign({ value: tags, border: "all", basis: "medium", placeholder: "No selection", onChange: onChangeTags }, props))),
        React.createElement(Box, { direction: "row", justify: "start" },
            React.createElement(Button, { primary: true, label: "Reset tags", onClick: resetTags }))));
};
export const main = () => React.createElement(TagsInput, null);
export const a11yTitle = () => React.createElement(TagsInput, { a11yTitle: "Grommet tags" });
export const children = () => {
    const [tags, setTags] = React.useState(defaultTags);
    const onChangeTags = ({ option }) => setTags(option);
    const removeTag = tagIndex => setTags(tags.filter((_, index) => index !== tagIndex));
    return (React.createElement(Tags, { value: tags, focusable: false, onChange: onChangeTags }, (tag, index) => (React.createElement(Box, { key: `remove_${index}`, pad: {
            horizontal: 'xsmall',
        } },
        React.createElement(Button, { label: tag, icon: React.createElement(Trash, null), onClick: () => removeTag(index) })))));
};
export const icon = () => React.createElement(TagsInput, { icon: React.createElement(FormSubtract, null) });
export const direction = () => React.createElement(TagsInput, { direction: "column" });
export const tagProps = () => (React.createElement(TagsInput, { tagProps: {
        background: 'status-critical',
        size: 'large',
        border: {
            color: 'brand',
            size: 'medium',
        },
    } }));
export const onClick = () => (React.createElement(TagsInput, { onClick: () => console.log('onClick') }));
export const focusable = () => React.createElement(TagsInput, { focusable: false });
export const placeholder = () => React.createElement(TagsInput, { placeholder: "No selection" });
export const value = () => React.createElement(TagsInput, { value: "tag" });
