// grommet-controls/dist/components/ColorInput/ColorInput.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { Component } from 'react';
import styled, { ThemeContext } from 'styled-components';
import { StopFill } from 'grommet-icons';
import { MaskedInput } from '../MaskedInput';
import { Colors } from '../Colors';
const StyledColorButton = styled.span `
  display: inline-flex;
  svg {
    border: ${props => `${props.theme.button.border.width} solid ${props.theme.global.colors.border}`};
    fill: ${props => props.selectedColor};
    stroke: ${props => props.selectedColor};
    transition: none;
  }
`;
/** A masked color entry field, with a drop button to select a color<br/>
 * All properties of MaskedInput apply<br/>
 * `import { ColorInput } from 'grommet-controls';`<br/>
 * `<ColorInput colors={...} value={...} />`<br/>
 */
class ColorInput extends Component {
    constructor(props) {
        super(props);
        this.upDateValue = null;
        this.onChange = e => {
            const { onChange } = this.props;
            this.setState({ selectedColor: e.target.value });
            if (onChange) {
                onChange(e);
            }
        };
        this.onSelect = ({ color }) => {
            this.setState({ selectedColor: color });
            this.upDateValue(color);
        };
        this.state = {
            selectedColor: (props.value || props.defaultValue),
        };
    }
    static getDerivedStateFromProps(nextProps) {
        return {
            selectedColor: nextProps.value,
        };
    }
    render() {
        const _a = this.props, { columns, wrap, colors, size } = _a, rest = __rest(_a, ["columns", "wrap", "colors", "size"]);
        const { selectedColor } = this.state;
        return (React.createElement(ThemeContext.Consumer, null, theme => (React.createElement(MaskedInput, Object.assign({ update: update => {
                this.upDateValue = update;
            }, dropIcon: React.createElement(StyledColorButton, { theme: theme, selectedColor: selectedColor },
                React.createElement(StopFill, null)), dropContent: React.createElement(Colors, { color: selectedColor, colors: colors, onSelect: this.onSelect, columns: columns, wrap: wrap, size: size }), onChange: this.onChange }, rest)))));
    }
}
ColorInput.defaultProps = {
    size: 'small',
    mask: [
        '#',
        /[0-9a-fA-F]/,
        /[0-9a-fA-F]/,
        /[0-9a-fA-F]/,
        /[0-9a-fA-F]/,
        /[0-9a-fA-F]/,
        /[0-9a-fA-F]/,
    ],
};
export { ColorInput };
