// grommet-controls/dist/components/Spinning/Spinning.js
// ported to styled-componnets from
// https://github.com/KyleAMathews/react-spinkit
import React from 'react';
import { ThemeContext } from 'styled-components';
import { allSpinners } from './spinners';
import { CircleSpinner, ThreeBounceSpinner, CubeGridSpinner, WaveSpinner, FoldingCubeSpinner, DoubleBounceSpinner, WanderingCubesSpinner, ChasingDotsSpinner, RotatingPlaneSpinner, PulseSpinner, WordPressSpinner, } from './StyledSpinning';
const styledComponents = {
    circle: CircleSpinner,
    'three-bounce': ThreeBounceSpinner,
    'cube-grid': CubeGridSpinner,
    wave: WaveSpinner,
    'folding-cube': FoldingCubeSpinner,
    'double-bounce': DoubleBounceSpinner,
    'wandering-cubes': WanderingCubesSpinner,
    'chasing-dots': ChasingDotsSpinner,
    'rotating-plane': RotatingPlaneSpinner,
    pulse: PulseSpinner,
    wordpress: WordPressSpinner,
};
/** A spinning process indicator<br/>
 * `import { Spinning } from 'grommet-controls';`<br/>
 * `<Spinning />`<br/>
 */
const Spinning = ({ kind, fadeIn, color, size }) => {
    const spinnerInfo = allSpinners[kind] || allSpinners['three-bounce'];
    const StyledSpinning = styledComponents[kind];
    return (React.createElement(ThemeContext.Consumer, null, theme => (React.createElement(StyledSpinning, { color: (color || (theme.dark ? 'light-1' : 'dark-1')), fadeIn: fadeIn, theme: theme, size: size }, Array.from(Array(spinnerInfo.divCount).keys()).map((_, idx) => (React.createElement("div", { key: idx })))))));
};
Spinning.defaultProps = {
    kind: 'circle',
    fadeIn: 'full',
    color: undefined,
    size: 'medium',
};
export { Spinning };
