// grommet-controls/dist/components/Spinning/Spinning.stories.js
import React from 'react';
import { Box } from 'grommet';
import { ControlTypes } from '@component-controls/core';
import { Spinning } from './Spinning';
export default {
    title: 'controls/Spinning',
    component: Spinning,
};
export const main = props => React.createElement(Spinning, Object.assign({}, props));
main.controls = {
    kind: {
        type: ControlTypes.OPTIONS,
        options: [
            'circle',
            'pulse',
            'three-bounce',
            'cube-grid',
            'wave',
            'folding-cube',
            'double-bounce',
            'wandering-cubes',
            'chasing-dots',
            'rotating-plane',
        ],
    },
    color: { type: ControlTypes.COLOR },
    size: {
        type: ControlTypes.OPTIONS,
        options: ['xsmall', 'small', 'medium', 'large', 'xlarge'],
    },
    fadeIn: {
        type: ControlTypes.OPTIONS,
        options: ['full', 'half', 'quarter', 'none', 'wordpress'],
    },
};
export const kind = () => (React.createElement(Box, { gap: "medium" },
    React.createElement(Spinning, { kind: "circle" }),
    React.createElement(Spinning, { kind: "pulse" }),
    React.createElement(Spinning, { kind: "three-bounce" }),
    React.createElement(Spinning, { kind: "cube-grid" }),
    React.createElement(Spinning, { kind: "wave" }),
    React.createElement(Spinning, { kind: "folding-cube" }),
    React.createElement(Spinning, { kind: "double-bounce" }),
    React.createElement(Spinning, { kind: "wandering-cubes" }),
    React.createElement(Spinning, { kind: "chasing-dots" }),
    React.createElement(Spinning, { kind: "rotating-plane" })));
export const color = () => (React.createElement(Box, { gap: "medium" },
    React.createElement(Spinning, { kind: "circle", color: "brand" }),
    React.createElement(Spinning, { kind: "pulse", color: "accent-1" }),
    React.createElement(Spinning, { kind: "three-bounce", color: "accent-2" }),
    React.createElement(Spinning, { kind: "cube-grid", color: "accent-3" }),
    React.createElement(Spinning, { kind: "wave", color: "neutral-1" }),
    React.createElement(Spinning, { kind: "folding-cube", color: "red" }),
    React.createElement(Spinning, { kind: "double-bounce", color: "blue" }),
    React.createElement(Spinning, { kind: "wandering-cubes" }),
    React.createElement(Spinning, { kind: "chasing-dots" }),
    React.createElement(Spinning, { kind: "rotating-plane" })));
export const size = () => (React.createElement(Box, { gap: "medium" },
    React.createElement(Spinning, { kind: "circle", color: "brand", size: "xsmall" }),
    React.createElement(Spinning, { kind: "pulse", color: "accent-1", size: "small" }),
    React.createElement(Spinning, { kind: "circle", color: "brand", size: "medium" }),
    React.createElement(Spinning, { kind: "three-bounce", color: "accent-2", size: "large" }),
    React.createElement(Spinning, { kind: "cube-grid", color: "accent-3", size: "xlarge" })));
