// grommet-controls/dist/components/uiColors/uiColors.js
export const aqua = '#00ffff';
export const azure = '#f0ffff';
export const beige = '#f5f5dc';
export const black = '#000000';
export const blue = '#0000ff';
export const brown = '#a52a2a';
export const cyan = '#00ffff';
export const darkblue = '#00008b';
export const darkcyan = '#008b8b';
export const darkgrey = '#a9a9a9';
export const darkgreen = '#006400';
export const darkkhaki = '#bdb76b';
export const darkmagenta = '#8b008b';
export const darkolivegreen = '#556b2f';
export const darkorange = '#ff8c00';
export const darkorchid = '#9932cc';
export const darkred = '#8b0000';
export const darksalmon = '#e9967a';
export const darkviolet = '#9400d3';
export const fuchsia = '#ff00ff';
export const gold = '#ffd700';
export const green = '#008000';
export const indigo = '#4b0082';
export const khaki = '#f0e68c';
export const lightblue = '#add8e6';
export const lightcyan = '#e0ffff';
export const lightgreen = '#90ee90';
export const lightgrey = '#d3d3d3';
export const lightpink = '#ffb6c1';
export const lightyellow = '#ffffe0';
export const lime = '#00ff00';
export const magenta = '#ff00ff';
export const maroon = '#800000';
export const navy = '#000080';
export const olive = '#808000';
export const orange = '#ffa500';
export const pink = '#ffc0cb';
export const purple = '#800080';
export const violet = '#800080';
export const red = '#ff0000';
export const silver = '#c0c0c0';
export const white = '#ffffff';
export const yellow = '#ffff00';
export const uiColors = {
    aqua,
    azure,
    beige,
    black,
    blue,
    brown,
    cyan,
    darkblue,
    darkcyan,
    darkgrey,
    darkgreen,
    darkkhaki,
    darkmagenta,
    darkolivegreen,
    darkorange,
    darkorchid,
    darkred,
    darksalmon,
    darkviolet,
    fuchsia,
    gold,
    green,
    indigo,
    khaki,
    lightblue,
    lightcyan,
    lightgreen,
    lightgrey,
    lightpink,
    lightyellow,
    lime,
    magenta,
    maroon,
    navy,
    olive,
    orange,
    pink,
    purple,
    violet,
    red,
    silver,
    white,
    yellow,
};
