// grommet-controls/dist/components/PagingTable/PagingTable.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { ThemeContext } from 'styled-components';
import { ResponsiveContext } from 'grommet';
import { StyledPagingTable } from './StyledPagingTable';
const responsiveColumns = (size, columns) => Array.isArray(columns)
    ? columns.map(column => {
        if (size === undefined) {
            return column;
        }
        let { show } = column;
        if (column.responsiveShow) {
            const sizes = Array.isArray(column.responsiveShow)
                ? column.responsiveShow
                : [column.responsiveShow];
            if (sizes.indexOf(size) !== -1) {
                show = true;
            }
        }
        if (column.responsiveHide) {
            const sizes = Array.isArray(column.responsiveHide)
                ? column.responsiveHide
                : [column.responsiveHide];
            if (sizes.indexOf(size) !== -1) {
                show = false;
            }
        }
        if (show === undefined) {
            return column;
        }
        return Object.assign(Object.assign({}, column), { show });
    })
    : columns;
/**
 * Table component with pagination, filtering, footer, child rows and grouping<br/>
 * Derived from https://github.com/react-tools/react-table<br/>
 * `import { PagingTable } from 'grommet-controls';`<br/>
 * `<PagingTable columns={...} data={...} />`<br/>
 */
const PagingTable = (_a) => {
    var { data, columns } = _a, rest = __rest(_a, ["data", "columns"]);
    const defaults = {
        showPagination: rest.onFetchData !== undefined ||
            (data && data.length > (rest.defaultPageSize || 20)),
        minRows: data && data.length < (rest.defaultPageSize || 20) ? 0 : undefined,
    };
    const newProps = Object.assign(Object.assign({}, defaults), rest);
    return (React.createElement(ResponsiveContext.Consumer, null, size => (React.createElement(ThemeContext.Consumer, null, theme => (React.createElement(StyledPagingTable, Object.assign({ data: data || [], columns: responsiveColumns(size, columns), theme: theme }, newProps)))))));
};
PagingTable.defaultProps = {
    defaultPageSize: 20,
    filterable: false,
    loading: false,
    manual: false,
    resizable: true,
    showPagination: true,
    sortable: true,
};
export { PagingTable };
