// grommet-controls/dist/components/Tag/Tag.stories.js
import React from 'react';
import { Github, Grommet, FormSubtract } from 'grommet-icons';
import { ControlTypes } from '@component-controls/core';
import { Tag } from './Tag';
export default {
    title: 'controls/Tag',
    component: Tag,
};
const tagIcons = [React.createElement(Github, { key: "github" }), React.createElement(Grommet, { key: "grommet" })];
export const main = props => React.createElement(Tag, Object.assign({}, props));
main.controls = {
    label: { type: ControlTypes.TEXT, value: 'Tag' },
    background: { type: ControlTypes.COLOR },
    size: {
        type: ControlTypes.OPTIONS,
        options: ['xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge'],
    },
    onChange: { type: ControlTypes.BUTTON, value: () => console.log('onChange') },
};
export const a11yTitle = () => React.createElement(Tag, { a11yTitle: "Tag component" });
export const icon = () => React.createElement(Tag, { icon: React.createElement(FormSubtract, null) });
export const disabled = () => React.createElement(Tag, { disabled: true, label: "Disabled" });
export const focusable = () => React.createElement(Tag, { focusable: false, label: "Cant focus" });
export const background = () => React.createElement(Tag, { background: "accent-1", label: "accent-1" });
export const border = () => (React.createElement(Tag, { border: {
        side: 'all',
        size: 'medium',
        color: 'accent-1',
    }, round: "medium", label: "border" }));
export const round = () => (React.createElement(Tag, { round: "medium", background: "accent-2", label: "round" }));
export const onClick = () => React.createElement(Tag, { onClick: () => console.log('onClick') });
export const onChange = () => {
    const [iconIndex, setIconIndex] = React.useState(0);
    const onToggle = () => {
        setIconIndex(1 - iconIndex);
    };
    return (React.createElement(Tag, { icon: tagIcons[iconIndex], onClick: () => console.log('onClick'), onChange: onToggle }));
};
export const reverse = () => React.createElement(Tag, { reverse: true, label: "Adam" });
export const size = () => React.createElement(Tag, { size: "large", label: "Large" });
