// grommet-controls/dist/components/Card/Card.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { ThemeContext } from 'styled-components';
import { Box, Heading } from 'grommet';
import { StyledCard } from './StyledCard';
/**
 * A Card-type container, all props of Box apply<br/>
 * `import { Card } from 'grommet-controls';`<br/>
 * `<Card>`<br/>
 * &nbsp;&nbsp;`<Card.CardTitle>`<br/>
 * &nbsp;&nbsp;&nbsp;&nbsp;`Title`<br/>
 * &nbsp;&nbsp;`</Card.CardTitle>`<br/>
 * &nbsp;&nbsp;`<Card.CardContent>`<br/>
 * &nbsp;&nbsp;&nbsp;&nbsp;`card content goes here`<br/>
 * &nbsp;&nbsp;`</Card.CardContent>`<br/>
 * `</Card>`<br/>
 */
export const Card = (_a) => {
    var { children } = _a, rest = __rest(_a, ["children"]);
    return (React.createElement(ThemeContext.Consumer, null, theme => (React.createElement(StyledCard, Object.assign({ theme: theme, overflow: "hidden" }, rest), children))));
};
Card.displayName = 'Card';
Card.defaultProps = {
    align: 'center',
    border: 'all',
    elevation: 'small',
    round: 'xsmall',
    flex: false,
    gap: 'small',
    fill: 'horizontal',
    background: 'white',
};
/**
 *  Card Actions to be placed at the bottom of the Card
 */
export const CardActions = (_a) => {
    var { children, pad = 'small' } = _a, rest = __rest(_a, ["children", "pad"]);
    return (React.createElement(Box, Object.assign({ align: "center", pad: pad, gap: "small", border: "top", flex: false, fill: "horizontal", direction: "row" }, rest), children));
};
CardActions.displayName = 'CardActions';
Card.CardActions = CardActions;
/**
 *  Card Title display the card title
 */
export const CardTitle = (_a) => {
    var { children, color, textAlign, truncate, level = 4, strong = false, responsive } = _a, rest = __rest(_a, ["children", "color", "textAlign", "truncate", "level", "strong", "responsive"]);
    return (React.createElement(Box, Object.assign({ direction: "row", fill: "horizontal", border: "bottom", gap: "small", flex: false }, rest), typeof children !== 'string' ? (children) : (React.createElement(Heading, { level: level, margin: "none", color: color, textAlign: textAlign, truncate: truncate, responsive: responsive }, strong ? React.createElement("strong", null, children) : children))));
};
CardTitle.displayName = 'CardTitle';
Card.CardTitle = CardTitle;
/**
 *  Card Content
 */
export const CardContent = (_a) => {
    var { children, pad = 'small' } = _a, rest = __rest(_a, ["children", "pad"]);
    return (React.createElement(Box, Object.assign({ pad: pad, fill: true, flex: true }, rest), children));
};
CardContent.displayName = 'CardContent';
Card.CardContent = CardContent;
