// grommet-controls/dist/components/Card/Card.stories.js
import React from 'react';
import { Box, Button, Heading, Image, Text } from 'grommet';
import { Add, Bitcoin, Edit } from 'grommet-icons';
import { ControlTypes } from '@component-controls/core';
import { Card } from './Card';
import { ImageStamp } from '../ImageStamp';
export default {
    title: 'controls/Card',
    component: Card,
};
const actions = [
    {
        label: 'Edit',
        icon: React.createElement(Edit, null),
        onClick: () => { },
    },
    {
        label: 'New',
        icon: React.createElement(Add, null),
        onClick: () => { },
    },
];
export const main = ({ border, align, direction, pad, imageStamp, title, level, fit, image, height, }) => (React.createElement(Card, null,
    React.createElement(Card.CardTitle, { border: border, align: align, direction: direction, pad: pad },
        React.createElement(Box, { direction: "row" },
            React.createElement(ImageStamp, Object.assign({}, imageStamp)),
            React.createElement(Heading, { level: level, margin: "xsmall" }, title)),
        React.createElement(Box, { direction: "row" },
            React.createElement(Bitcoin, { color: "plain" }),
            React.createElement(Text, null, "sub-title info"))),
    React.createElement(Card.CardContent, null,
        React.createElement(Image, { fit: fit, src: image, height: `${height}` })),
    React.createElement(Card.CardActions, { justify: "center" }, actions.map((action, index) => (React.createElement(Button, { key: `actions_${index}`, label: action.label, onClick: action.onClick }))))));
main.controls = {
    border: {
        type: 'object',
        value: {
            color: {
                type: ControlTypes.COLOR,
                value: '#333',
            },
            size: {
                type: ControlTypes.OPTIONS,
                options: ['xsmall', 'small', 'medium', 'large', 'xlarge'],
            },
            side: {
                type: ControlTypes.OPTIONS,
                options: [
                    'top',
                    'left',
                    'bottom',
                    'right',
                    'horizontal',
                    'vertical',
                    'all',
                ],
                value: 'bottom',
            },
        },
        groupId: 'Card.CardTitle',
    },
    align: {
        type: ControlTypes.OPTIONS,
        options: ['start', 'center', 'end', 'baseline', 'stretch'],
        value: 'center',
        groupId: 'Card.CardTitle',
    },
    direction: {
        type: ControlTypes.OPTIONS,
        options: [
            'row',
            'column',
            'row-responsive',
            'row-reverse',
            'column-reverse',
        ],
        value: 'column',
        groupId: 'Card.CardTitle',
    },
    pad: {
        type: ControlTypes.OPTIONS,
        options: [
            'none',
            'xxsmall',
            'xsmall',
            'small',
            'medium',
            'large',
            'xlarge',
        ],
        value: 'small',
        groupId: 'Card.CardTitle',
    },
    imageStamp: {
        type: ControlTypes.OBJECT,
        value: {
            src: {
                type: ControlTypes.TEXT,
                value: 'https://s3.amazonaws.com/uifaces/faces/twitter/shadeed9/128.jpg',
            },
            round: {
                type: ControlTypes.OPTIONS,
                options: ['xsmall', 'small', 'medium', 'large', 'full'],
                value: 'full',
            },
        },
        groupId: 'Heading',
    },
    title: {
        type: ControlTypes.TEXT,
        value: 'Info card',
        groupId: 'Heading',
    },
    level: {
        type: ControlTypes.NUMBER,
        value: 3,
        min: 1,
        max: 5,
        groupId: 'Heading',
    },
    fit: {
        type: ControlTypes.OPTIONS,
        value: 'cover',
        options: ['cover', 'contain'],
        groupId: 'Content',
        display: 'inline-radio',
    },
    image: {
        type: ControlTypes.TEXT,
        value: 'https://s3.amazonaws.com/uifaces/faces/twitter/denisepires/128.jpg',
        groupId: 'Content',
        data: { name: 'internet.avatar' },
    },
    height: {
        type: ControlTypes.NUMBER,
        value: 250,
        min: 100,
        max: 900,
        step: 50,
        range: true,
        groupId: 'Content',
    },
};
export const animation = () => (React.createElement(Card, { animation: {
        type: 'zoomIn',
        duration: 5000,
        size: 'xlarge',
    } },
    React.createElement(Card.CardTitle, { border: "bottom" }, "Card"),
    React.createElement(Card.CardContent, null,
        React.createElement(Image, { fit: "cover", src: "//v2.grommet.io/assets/Wilderpeople_Ricky.jpg", height: "250" }))));
export const background = () => (React.createElement(Card, { background: "accent-1" },
    React.createElement(Card.CardTitle, { border: "bottom" }, "Card"),
    React.createElement(Card.CardContent, { basis: "small" }, "Card content")));
export const border = () => (React.createElement(Card, { border: {
        color: 'brand',
        size: 'medium',
    } },
    React.createElement(Card.CardTitle, { border: "bottom" }, "Card"),
    React.createElement(Card.CardContent, { basis: "small" }, "Card content")));
export const elevation = () => (React.createElement(Card, { elevation: "large" },
    React.createElement(Card.CardTitle, { border: "bottom" }, "Card"),
    React.createElement(Card.CardContent, { basis: "small" }, "Card content")));
export const gap = () => (React.createElement(Card, { gap: "medium" },
    React.createElement(Card.CardTitle, { border: "bottom" }, "Card"),
    React.createElement(Card.CardContent, { basis: "small" }, "Card content")));
export const pad = () => (React.createElement(Card, { pad: "medium" },
    React.createElement(Card.CardTitle, { border: "bottom" }, "Card"),
    React.createElement(Card.CardContent, { basis: "small" }, "Card content")));
export const round = () => (React.createElement(Card, { round: "medium", pad: "small" },
    React.createElement(Card.CardTitle, { border: "bottom" }, "Card"),
    React.createElement(Card.CardContent, { basis: "small" }, "Card content")));
