// grommet-controls/dist/components/MaskedInput/MaskedInput.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import { createTextMaskInputElement } from 'text-mask-core';
import { createAutoCorrectedDatePipe, emailMask, createNumberMask, } from 'text-mask-addons';
import { DropInput } from '../DropInput';
import { transformMaskedValue } from './utils';
import { createMinMaxInputPipe, maskedNumberValue } from './minMaxNumberPipe';
/** A masked Input control with an optional drop button with the specified dropContent<br/>
 * `import { MakedInput } from 'grommet-controls';`<br/>
 * `<MakedInput mask={...} />`<br/>
 */
class MaskedInput extends Component {
    constructor() {
        super(...arguments);
        this.textMaskInputElement = null;
        this.inputControlRef = React.createRef();
        this.onInput = event => {
            const { onInput } = this.props;
            if (this.textMaskInputElement) {
                this.textMaskInputElement.update(event.target.value);
            }
            if (onInput) {
                onInput(event);
            }
        };
    }
    initTextMask() {
        const { props, props: { value }, } = this;
        if (props.mask && this.inputControlRef) {
            this.textMaskInputElement = createTextMaskInputElement(Object.assign({ 
                // eslint-disable-next-line react/no-find-dom-node
                inputElement: findDOMNode(this.inputControlRef.current).getElementsByTagName('input')[0] }, props));
            this.textMaskInputElement.update(value);
        }
    }
    componentDidMount() {
        this.initTextMask();
    }
    componentDidUpdate() {
        this.initTextMask();
    }
    maskedValue(value) {
        if (value === undefined) {
            return undefined;
        }
        const { mask } = this.props;
        if (value) {
            const conformed = transformMaskedValue(value, mask, this.props);
            return conformed;
        }
        return value;
    }
    render() {
        const _a = this.props, { defaultValue, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        value, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        mask, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        guide, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        showMask, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        pipe, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        placeholderChar, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        keepCharPositions } = _a, rest = __rest(_a, ["defaultValue", "value", "mask", "guide", "showMask", "pipe", "placeholderChar", "keepCharPositions"]);
        return (React.createElement(DropInput, Object.assign({ ref: this.inputControlRef, value: this.maskedValue(value), defaultValue: this.maskedValue(defaultValue), onInput: this.onInput }, rest)));
    }
}
MaskedInput.defaultProps = {
    type: 'text',
    guide: true,
    showMask: false,
    placeholderChar: '\u2000',
};
MaskedInput.alphabetic = /[A-Z]/i;
MaskedInput.placeholderChars = {
    whitespace: '\u2000',
    underscore: '_',
};
MaskedInput.digit = /\d/;
MaskedInput.createAutoCorrectedDatePipe = createAutoCorrectedDatePipe;
MaskedInput.emailMask = emailMask;
MaskedInput.createNumberMask = createNumberMask;
MaskedInput.createMinMaxInputPipe = createMinMaxInputPipe;
MaskedInput.maskedNumberValue = maskedNumberValue;
export { MaskedInput };
