// grommet-controls/dist/components/MaskedInput/MaskedInput.stories.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { Box, Calendar } from 'grommet';
import { Add, Subtract } from 'grommet-icons';
import { MaskedInput } from './MaskedInput';
import { smallDate } from '../../utils/moment';
export default {
    title: 'input/MaskedInput',
    component: MaskedInput,
};
const MaskedTestBed = (_a) => {
    var { value: defaultValue = '' } = _a, props = __rest(_a, ["value"]);
    const [value, setValue] = React.useState(defaultValue);
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium", gap: "small" },
            React.createElement(MaskedInput, Object.assign({ value: value, onChange: ({ target }) => setValue(target.value) }, props)))));
};
const DateInput = props => {
    const [date, setDate] = React.useState(smallDate(new Date('08/30/2020')));
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium" },
            React.createElement(MaskedInput, Object.assign({ mask: [/\d/, /\d/, '/', /\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/], dropContent: React.createElement(Box, { pad: "small" },
                    React.createElement(Calendar, { size: "small", date: date, onSelect: isoDate => setDate(smallDate(new Date(isoDate))) })), value: date, onChange: ({ target }) => setDate(target.value) }, props)))));
};
const NumberInput = () => {
    const [number, setNumber] = React.useState(10);
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium" },
            React.createElement(MaskedInput, { value: number, onChange: ({ target: { value } }) => setNumber(parseFloat(value)), mask: MaskedInput.createNumberMask({
                    allowDecimal: true,
                }), widgets: [
                    {
                        icon: React.createElement(Add, null),
                        onClick: () => setNumber(number + 1),
                    },
                    {
                        icon: React.createElement(Subtract, null),
                        onClick: () => setNumber(number - 1),
                    },
                ] }))));
};
export const main = () => (React.createElement(MaskedTestBed, { value: "8024442131", mask: [
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ] }));
export const a11yTitle = () => (React.createElement(MaskedTestBed, { a11yTitle: "Dollars", mask: MaskedInput.createNumberMask({}), value: "18933.85" }));
export const disabled = () => React.createElement(DateInput, { disabled: true });
export const dropContent = () => React.createElement(DateInput, null);
export const widgets = () => React.createElement(NumberInput, null);
export const plain = () => (React.createElement(MaskedTestBed, { value: "8024442131", mask: [
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ], plain: true }));
export const focusIndicator = () => (React.createElement(MaskedTestBed, { value: "8024442131", mask: [
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ], plain: true, focusIndicator: true }));
export const mask = () => (React.createElement(MaskedTestBed, { placeholder: "US phone number with country code", mask: [
        '+',
        '1',
        ' ',
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ] }));
export const guide = () => (React.createElement(MaskedTestBed, { value: "8024442131", guide: false, mask: [
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ], plain: true, focusIndicator: true }));
export const pipe = () => (React.createElement(MaskedTestBed, { mask: [
        MaskedInput.alphabetic,
        MaskedInput.digit,
        MaskedInput.alphabetic,
        ' ',
        MaskedInput.digit,
        MaskedInput.alphabetic,
        MaskedInput.digit,
    ], pipe: conformedValue => ({
        value: conformedValue.toUpperCase(),
    }), placeholder: "K1A 0B2", placeholderChar: MaskedInput.placeholderChars.underscore }));
export const placeholderChar = () => (React.createElement(MaskedTestBed, { mask: [
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ], placeholderChar: MaskedInput.placeholderChars.underscore, value: "8024442131" }));
export const keepCharPositions = () => (React.createElement(MaskedTestBed, { mask: [
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ], keepCharPositions: true, value: "8024442131" }));
export const showMask = () => (React.createElement(MaskedTestBed, { mask: [
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ], showMask: true, value: "8024442131" }));
export const placeholder = () => (React.createElement(MaskedTestBed, { mask: [
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ], placeholder: "Enter phone...", value: "8024442131" }));
export const name = () => (React.createElement(MaskedTestBed, { mask: [
        '(',
        /[1-9]/,
        /\d/,
        /\d/,
        ')',
        ' ',
        /\d/,
        /\d/,
        /\d/,
        '-',
        /\d/,
        /\d/,
        /\d/,
        /\d/,
    ], id: "date-id", name: "date-name", value: "8024442131" }));
