// grommet-controls/dist/components/Colors/Colors.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import { ThemeContext } from 'styled-components';
import { Box, Keyboard } from 'grommet';
import { normalizeColor } from 'grommet/utils/colors';
import { parseMetricToNum } from 'grommet/utils/mixins';
import { StyledColors, StyledColor, StyledColorContainer, StyledRow, StyledRows, } from './StyledColors';
const buildState = (props) => {
    const { colors, color, columns, wrap } = props;
    let colorRows = [];
    let lastRow = null;
    if (colors) {
        Object.keys(colors).forEach(item => {
            if (typeof colors[item] === 'object') {
                const row = { name: item, colors: [], buttonRef: React.createRef() };
                Object.keys(colors[item]).forEach(colorName => {
                    const colorsColor = colors[item][colorName];
                    row.colors.push({ name: colorName, color: colorsColor });
                });
                colorRows.push(row);
            }
            else {
                if (!lastRow) {
                    lastRow = { name: undefined, colors: [] };
                    colorRows.push(lastRow);
                }
                lastRow.colors.push({
                    name: item,
                    color: colors[item],
                    buttonRef: React.createRef(),
                });
            }
        });
    }
    const colorsPerRow = columns ||
        (colorRows.length === 1
            ? Math.sqrt(colorRows[0].colors.length)
            : colorRows.reduce((res, row) => Math.min(res, row.colors.length), Number.MAX_SAFE_INTEGER));
    let wrapColors = wrap;
    if (wrapColors === undefined) {
        wrapColors = columns === undefined && colorRows.length === 1;
    }
    if (wrapColors) {
        const finalRows = [];
        colorRows.forEach(row => {
            while (row.colors.length > 0) {
                finalRows.push({
                    name: row.name,
                    colors: row.colors.splice(0, colorsPerRow),
                });
            }
        });
        colorRows = finalRows;
    }
    else {
        colorRows = colorRows.map(row => (Object.assign(Object.assign({}, row), { colors: row.colors.filter((_, index) => index < colorsPerRow) })));
    }
    let activeRow = 0;
    let activeColor = 0;
    colorRows.find((row, rowIndex) => row.colors.find((item, colorIndex) => {
        const found = item.color === color;
        if (found) {
            activeRow = rowIndex;
            activeColor = colorIndex;
        }
        return found;
    }));
    return { colorRows, activeRow, activeColor };
};
/** A color selection element, with custom color palettes.<br/>
 *`import { Colors } from 'grommet-controls';`<br/>
 *`<Colors colors={...} />`<br/>
 */
class Colors extends Component {
    constructor(props) {
        super(props);
        this.setFocus = false;
        this.timer = 0;
        this.setActive = ({ rowIndex, colorIndex }) => {
            const { colorRows } = this.state;
            this.setFocus = true;
            const activeRow = Math.min(Math.max(0, rowIndex), colorRows.length - 1);
            const color = Math.max(0, colorIndex);
            const activeColor = Math.min(color, colorRows[activeRow].colors.length - 1);
            this.setState({
                activeRow,
                activeColor,
            });
        };
        this.onClickColor = ({ color, rowIndex, colorIndex, rowName, colorName, }) => () => {
            const { onSelect } = this.props;
            this.setActive({ rowIndex, colorIndex });
            if (onSelect) {
                onSelect({ color, rowName, colorName });
            }
        };
        this.state = buildState(props);
    }
    static getDerivedStateFromProps(nextProps) {
        return buildState(nextProps);
    }
    componentDidUpdate() {
        if (this.setFocus) {
            const { activeRow, activeColor, colorRows } = this.state;
            this.setFocus = false;
            if (colorRows[activeRow] &&
                colorRows[activeRow].colors[activeColor] &&
                colorRows[activeRow].colors[activeColor].buttonRef) {
                // eslint-disable-next-line react/no-find-dom-node
                const buttonNode = findDOMNode(colorRows[activeRow].colors[activeColor].buttonRef.current);
                if (buttonNode) {
                    buttonNode.scrollIntoView();
                    buttonNode.focus();
                }
            }
        }
    }
    componentWillUnmount() {
        clearTimeout(this.timer);
    }
    render() {
        const _a = this.props, { 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        wrap, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onSelect, 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        columns, size } = _a, rest = __rest(_a, ["wrap", "onSelect", "columns", "size"]);
        const { colorRows, activeRow, activeColor } = this.state;
        const colors = (theme, cellSize) => colorRows.map((row, rowIndex) => (React.createElement(StyledRow, { key: `row_${rowIndex}`, theme: theme, style: {
                width: `${cellSize * row.colors.length}px`,
                height: `${cellSize}px`,
            } }, row.colors.map((color, colorIndex) => {
            const isActive = activeRow === rowIndex && activeColor === colorIndex;
            const colorStyle = {
                backgroundColor: color.color,
                color: normalizeColor('text', theme),
                left: `${cellSize * colorIndex}px`,
            };
            return (React.createElement(StyledColorContainer, { key: `color_${colorIndex}`, theme: theme },
                React.createElement(StyledColor
                // eslint-disable-next-line no-param-reassign
                , { 
                    // eslint-disable-next-line no-param-reassign
                    ref: color.buttonRef, style: colorStyle, size: size, theme: theme, tabIndex: isActive ? 0 : -1, a11yTitle: `${row.name !== undefined ? row.name : ''} ${color.name}`, plain: true, active: isActive, hoverIndicator: "background", onClick: this.onClickColor({
                        color: color.color,
                        rowIndex,
                        colorIndex,
                        colorName: color.name,
                        rowName: row.name,
                    }) },
                    React.createElement("span", null, color.color))));
        }))));
        return (React.createElement(ThemeContext.Consumer, null, theme => {
            const cellSize = parseMetricToNum(theme.calendar[size].daySize);
            return (React.createElement(StyledColors, Object.assign({ size: size, theme: theme }, rest),
                React.createElement(Keyboard, { onUp: event => {
                        event.preventDefault();
                        this.setActive({
                            rowIndex: activeRow - 1,
                            colorIndex: activeColor,
                        });
                    }, onDown: event => {
                        event.preventDefault();
                        this.setActive({
                            rowIndex: activeRow + 1,
                            colorIndex: activeColor,
                        });
                    }, onLeft: () => this.setActive({
                        rowIndex: activeRow,
                        colorIndex: activeColor - 1,
                    }), onRight: () => this.setActive({
                        rowIndex: activeRow,
                        colorIndex: activeColor + 1,
                    }) },
                    React.createElement(Box, null,
                        React.createElement(StyledRows, { style: { height: `${cellSize * colorRows.length}px` } }, colors(theme, cellSize))))));
        }));
    }
}
Colors.defaultProps = {
    size: 'medium',
};
export { Colors };
