// grommet-controls/dist/components/PasswordInput/PasswordInput.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { Component } from 'react';
import { FormView } from 'grommet-icons/icons/FormView';
import { Hide } from 'grommet-icons/icons/Hide';
import { DropInput } from '../DropInput';
/**
 *A password field with show/hide password toggle<br/>
 *All properties of DropInput apply<br/>
 *`import { PasswordInput } from 'grommet-controls';`<br/>
 *`<PasswordInput value={...} />`<br/>
 */
class PasswordInput extends Component {
    constructor() {
        super(...arguments);
        this.state = { showPassword: false };
        this.toggleView = e => {
            const { showPassword } = this.state;
            e.stopPropagation();
            this.setState({ showPassword: !showPassword });
        };
    }
    render() {
        const _a = this.props, { viewIcon, hideIcon, a11yViewPassword, a11yHidePassword } = _a, rest = __rest(_a, ["viewIcon", "hideIcon", "a11yViewPassword", "a11yHidePassword"]);
        const { showPassword } = this.state;
        return (React.createElement(DropInput, Object.assign({ type: showPassword ? 'text' : 'password', widgets: [
                {
                    icon: showPassword ? hideIcon : viewIcon,
                    'aria-label': showPassword ? a11yHidePassword : a11yViewPassword,
                    onClick: this.toggleView,
                },
            ] }, rest)));
    }
}
PasswordInput.defaultProps = {
    viewIcon: React.createElement(FormView, null),
    hideIcon: React.createElement(Hide, null),
    a11yViewPassword: 'View password',
    a11yHidePassword: 'Hide password',
};
export { PasswordInput };
