// grommet-controls/dist/components/Header/Header.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { StyledHeader } from './StyledHeader';
/**
 * A page header component
 * ```$ npm install grommet-controls
 * import { Header } from 'grommet-controls';
 * <Header>
 *   ...
 * </Header>
 * ```
 */
const Header = (_a) => {
    var { children } = _a, rest = __rest(_a, ["children"]);
    return (React.createElement(StyledHeader, Object.assign({ tag: "header" }, rest), children));
};
Header.defaultProps = {
    position: 'relative',
    zIndex: 10,
    direction: 'row',
    background: 'brand',
    fill: 'horizontal',
    justify: 'between',
    align: 'center',
    elevation: 'medium',
    pad: 'medium',
};
export { Header };
