// grommet-controls/dist/index.esm2.js
import { __rest } from 'tslib';
import React from 'react';
import { ThemeContext } from 'grommet';
import { deepMerge } from 'grommet/utils/object';
import { normalizeColor, getRGBA } from 'grommet/utils/colors';
import { a as colorForName, c as colorFromIndex } from './colors.esm.js';
import { Bar, Doughnut, HorizontalBar, Line, Pie, Polar, Radar, Scatter } from 'react-chartjs-2';

const normalizedColor = (color, theme) => colorForName(normalizeColor(color, theme), theme);
const withChartTheme = (WrappedComponent, _a = {}) => {
    var { classOpacity = 0.6, defaultScales = [{}] } = _a, other = __rest(_a, ["classOpacity", "defaultScales"]);
    return ({ options, data }) => (React.createElement(ThemeContext.Consumer, null, theme => {
        const textColor = normalizedColor('text', theme);
        const axisColors = {
            ticks: {
                fontColor: textColor,
            },
            gridLines: {
                color: normalizedColor('border', theme),
            },
            scaleLabel: {
                fontColor: textColor,
            },
        };
        const defaultOptions = {
            maintainAspectRatio: false,
            title: {
                fontColor: textColor,
            },
            legend: {
                labels: {
                    fontColor: textColor,
                },
            },
            scales: {},
        };
        const themedOptions = deepMerge(defaultOptions, options);
        themedOptions.scales.xAxes = themedOptions.scales.xAxes || defaultScales;
        themedOptions.scales.yAxes = themedOptions.scales.yAxes || defaultScales;
        themedOptions.scales.xAxes = themedOptions.scales.xAxes.map(x => deepMerge(axisColors, x));
        themedOptions.scales.yAxes = themedOptions.scales.yAxes.map(y => deepMerge(axisColors, y));
        let datasets;
        if (data && Array.isArray(data.datasets)) {
            datasets = data.datasets.map((dataset, i) => {
                const { backgroundColor, borderColor, color } = dataset, rest = __rest(dataset, ["backgroundColor", "borderColor", "color"]);
                let newOpts;
                const themeColors = (index, itemOpacity) => {
                    const lineColor = borderColor || color || colorFromIndex(index);
                    const lineColors = Array.isArray(lineColor)
                        ? lineColor.map(c => normalizedColor(c, theme))
                        : normalizedColor(lineColor, theme);
                    const fillColor = backgroundColor || lineColor;
                    const opacity = itemOpacity ||
                        dataset.opacity ||
                        (options && options.opacity) ||
                        classOpacity;
                    const fillColors = Array.isArray(fillColor)
                        ? fillColor.map(c => getRGBA(normalizedColor(c, theme), opacity))
                        : getRGBA(normalizedColor(fillColor, theme), opacity);
                    return {
                        backgroundColor: fillColors,
                        borderColor: lineColors,
                    };
                };
                if (options && options.themedData && dataset.data) {
                    newOpts = { backgroundColor: [], borderColor: [] };
                    dataset.data.forEach((dataRow, rIndex) => {
                        const colors = themeColors(rIndex, dataRow.opacity);
                        newOpts.backgroundColor.push(colors.backgroundColor);
                        newOpts.borderColor.push(colors.borderColor);
                    });
                }
                else {
                    newOpts = themeColors(i);
                }
                return Object.assign(Object.assign(Object.assign({}, newOpts), other), rest);
            });
        }
        return (React.createElement(WrappedComponent, { options: themedOptions, data: Object.assign(Object.assign({}, data), { datasets }) }));
    }));
};

/**
 * Bar chart wrapper to chart.js Bar with grommet theming<br/>
 * For all standard charting options, consult the chart.js docs<br/>
 * `import { BarChart } from 'grommet-controls/chartjs';`<br/>
 * `<BarChart data={...} options={...} />`<br/>
 */
const BarChart = withChartTheme(Bar, { borderWidth: 3 });

/** Doughnut chart wrapper to chart.js Doughnut with grommet theming,br/>
 * For all standard charting options, consult the chart.js docs<br/>
 * `import { DoughnutChart } from 'grommet-controls/chartjs';`<br/>
 * `<DoughnutChart data={...} options={...} />`<br/>
 */
const DoughnutChart = withChartTheme(Doughnut, {
    classOpacity: 1,
    defaultScales: [],
});

/** HorizontalBar chart wrapper to chart.js HorizontalBar with grommet theming<br/>
 * For all standard charting options, consult the chart.js docs<br/>
 * `import { HorizontalBarChart } from 'grommet-controls/chartjs';`<br/>
 * `<HorizontalBarChart data={...} options={...} />`<br/>
 */
const HorizontalBarChart = withChartTheme(HorizontalBar, { borderWidth: 3 });

/** Line chart wrapper to chart.js Line with grommet theming<br/>
 * For all standard charting options, consult the chart.js docs<br/>
 * `import { LineChart } from 'grommet-controls/chartjs';`<br/>
 * `<LineChart data={...} options={...} />`<br/>
 */
const LineChart = withChartTheme(Line);

/** Pie chart wrapper to chart.js Pie with grommet theming<br/>
 * For all standard charting options, consult the chart.js docs<br/>
 * `import { PieChart } from 'grommet-controls/chartjs';`<br/>
 * `<PieChart data={...} options={...} />`<br/>
 */
const PieChart = withChartTheme(Pie, { classOpacity: 1, defaultScales: [] });

/** Polar chart wrapper to chart.js Polar with grommet theming<br/>
 * For all standard charting options, consult the chart.js docs<br/>
 * `import { PolarChart } from 'grommet-controls/chartjs';`<br/>
 * `<PolarChart data={...} options={...} />`<br/>
 */
const PolarChart = withChartTheme(Polar, { defaultScales: [] });

/** Radar chart wrapper to chart.js Radar with grommet theming<br/>
 * For all standard charting options, consult the chart.js docs<br/>
 * `import { RadarChart } from 'grommet-controls/chartjs';`<br/>
 * `<RadarChart data={...} options={...} />`<br/>
 */
const RadarChart = withChartTheme(Radar, { defaultScales: [] });

/** Scatter chart wrapper to chart.js Scatter with grommet theming<br/>
 * For all standard charting options, consult the chart.js docs<br/>
 * `import { ScatterChart } from 'grommet-controls/chartjs';`<br/>
 * `<ScatterChart data={...} options={...} />`<br/>
 */
const ScatterChart = withChartTheme(Scatter);

export { BarChart, DoughnutChart, HorizontalBarChart, LineChart, PieChart, PolarChart, RadarChart, ScatterChart, withChartTheme };
