// grommet-controls/dist/components/NumberInput/NumberInput.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { Component } from 'react';
import { Add } from 'grommet-icons/icons/Add';
import { Subtract } from 'grommet-icons/icons/Subtract';
import { MaskedInput } from '../MaskedInput';
const precision = n => {
    // eslint-disable-next-line no-restricted-globals
    if (!isFinite(n))
        return 0;
    let e = 1;
    let p = 0;
    while (Math.round(n * e) / e !== n) {
        e *= 10;
        p += 1;
    }
    return p;
};
const minMax = ({ value, min, max }) => {
    let val = value;
    if (min !== undefined) {
        val = Math.max(val, min);
    }
    if (max !== undefined) {
        val = Math.min(val, max);
    }
    return val;
};
/** A masked number input, with widgets to increment/decrement the value<br/>
 *  Additionally, all properties of MaskedInput apply<br/>
 *  `import { NumberInput } from 'grommet-controls';`<br/>
 *  `<NumberInput value={...} />`<br/>
 */
class NumberInput extends Component {
    constructor() {
        super(...arguments);
        this.upDateValue = null;
        this.value = undefined;
        this.valueToNumber = (value) => {
            const { prefix, suffix, thousandsSeparatorSymbol, decimalSymbol, } = this.props;
            return MaskedInput.maskedNumberValue({
                value,
                prefix,
                suffix,
                thousandsSeparatorSymbol,
                decimalSymbol,
            });
        };
        this.isFloat = (val) => {
            const floatRegex = /^-?\d+(?:[.,]\d*?)?$/;
            if (!floatRegex.test(val.toString())) {
                return false;
            }
            // eslint-disable-next-line no-restricted-globals
            if (isNaN(parseFloat(val.toString()))) {
                return false;
            }
            return true;
        };
        this.addStep = () => {
            const { max, min, step, value } = this.props;
            let val = this.isFloat(value)
                ? (this.valueToNumber(value) + step).toFixed(precision(step))
                : min || 0;
            if (typeof val === 'number' && Number.isNaN(val)) {
                if (min !== undefined) {
                    val = min;
                }
                else {
                    val = undefined;
                }
            }
            else {
                val = minMax({ value: val, min, max });
            }
            this.upDateValue(val.toString());
        };
        this.subtractStep = () => {
            const { max, min, step, value } = this.props;
            let val = this.isFloat(value)
                ? (this.valueToNumber(value) - step).toFixed(precision(step))
                : max || 0;
            if (typeof val === 'number' && Number.isNaN(val)) {
                if (max !== undefined) {
                    val = max;
                }
                else {
                    val = undefined;
                }
            }
            else {
                val = minMax({ value: val, min, max });
            }
            this.upDateValue(val.toString());
        };
        this.onChange = e => {
            const { onChange, emptyValue } = this.props;
            if (onChange) {
                let { value } = e.target;
                if (this.value !== value) {
                    if (value === undefined) {
                        value = emptyValue;
                    }
                    this.value = value;
                    onChange(Object.assign(Object.assign({}, e), { target: Object.assign(Object.assign({}, e.target), { value }) }));
                }
            }
        };
    }
    render() {
        const _a = this.props, { 
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onChange, min, max, step, pipe: userPipe, prefix, suffix, thousandsSeparatorSymbol, decimalSymbol, decimals, integers, a11yIncrement, a11yDecrement, mask: userMask, addIcon, subtractIcon, disabled } = _a, rest = __rest(_a, ["onChange", "min", "max", "step", "pipe", "prefix", "suffix", "thousandsSeparatorSymbol", "decimalSymbol", "decimals", "integers", "a11yIncrement", "a11yDecrement", "mask", "addIcon", "subtractIcon", "disabled"]);
        const allowNegative = typeof min !== 'number' || min < 0;
        const includeThousandsSeparator = !!thousandsSeparatorSymbol;
        const allowDecimal = decimals === null || (typeof decimals === 'number' && decimals > 0);
        const mask = userMask ||
            MaskedInput.createNumberMask({
                prefix,
                suffix,
                includeThousandsSeparator,
                thousandsSeparatorSymbol,
                allowDecimal,
                decimalSymbol,
                decimalLimit: decimals,
                integerLimit: integers,
                allowNegative,
            });
        const pipe = userPipe ||
            MaskedInput.createMinMaxInputPipe(Object.assign({ mask,
                prefix,
                suffix,
                thousandsSeparatorSymbol,
                decimalSymbol,
                min,
                max }, rest));
        return (React.createElement(MaskedInput, Object.assign({ update: update => {
                this.upDateValue = update;
            }, pattern: "[0-9]*", inputMode: "numeric", onKeyDown: this.subtractStep, onKeyUp: this.addStep, disabled: disabled, onChange: this.onChange, pipe: pipe, mask: mask, widgets: [
                {
                    icon: addIcon,
                    onClick: disabled ? undefined : this.addStep,
                    'aria-label': `${a11yIncrement} ${step}`,
                },
                {
                    icon: subtractIcon,
                    onClick: disabled ? undefined : this.subtractStep,
                    'aria-label': `${a11yDecrement} ${step}`,
                },
            ] }, rest)));
    }
}
NumberInput.defaultProps = {
    emptyValue: '',
    step: 1,
    addIcon: React.createElement(Add, null),
    subtractIcon: React.createElement(Subtract, null),
    prefix: '',
    suffix: '',
    thousandsSeparatorSymbol: '',
    decimalSymbol: '.',
    decimals: null,
    integers: null,
    a11yIncrement: 'Increment by',
    a11yDecrement: 'Decrement by',
};
export { NumberInput };
