// grommet-controls/dist/components/NumberInput/NumberInput.stories.js
import React from 'react';
import { Box } from 'grommet';
import { AddCircle, SubtractCircle } from 'grommet-icons';
import { NumberInput } from './NumberInput';
export default {
    title: 'input/NumberInput',
    component: NumberInput,
};
export const main = () => {
    const [number, setNumber] = React.useState('12000.34');
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium", gap: "small" },
            React.createElement(NumberInput, { value: number, thousandsSeparatorSymbol: ",", onChange: ({ target: { value } }) => setNumber(value) }))));
};
const IntegerInput = props => {
    const [number, setNumber] = React.useState('5');
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium", gap: "small" },
            React.createElement(NumberInput, Object.assign({ value: number, step: 1, onChange: ({ target: { value } }) => setNumber(value) }, props)))));
};
const FloatInput = props => {
    const [number, setNumber] = React.useState('2000.23');
    return (React.createElement(Box, { direction: "row" },
        React.createElement(Box, { basis: "medium", gap: "small" },
            React.createElement(NumberInput, Object.assign({ value: number, step: 1, onChange: ({ target: { value } }) => setNumber(value) }, props)))));
};
export const a11yIncrement = () => (React.createElement(IntegerInput, { a11yIncrement: "Add to entry", step: 10, a11yDecrement: "Remove from entry", thousandsSeparatorSymbol: "," }));
export const min = () => React.createElement(IntegerInput, { min: 5 });
export const max = () => React.createElement(IntegerInput, { max: 5 });
export const step = () => React.createElement(IntegerInput, { step: 10 });
export const disabled = () => React.createElement(IntegerInput, { disabled: true });
export const addIcon = () => (React.createElement(IntegerInput, { addIcon: React.createElement(AddCircle, null), subtractIcon: React.createElement(SubtractCircle, null) }));
export const prefix = () => React.createElement(IntegerInput, { prefix: "$" });
export const suffix = () => React.createElement(IntegerInput, { suffix: " USD" });
export const thousandsSeparatorSymbol = () => (React.createElement(FloatInput, { thousandsSeparatorSymbol: " ", decimalSymbol: ",", updateToString: true }));
export const decimalSymbol = () => (React.createElement(FloatInput, { thousandsSeparatorSymbol: "", decimalSymbol: ",", updateToString: true }));
export const decimals = () => React.createElement(FloatInput, { decimals: 4 });
export const integers = () => React.createElement(IntegerInput, { integers: 4 });
